lib/fog/vsphere/models/compute/server.rb

```diff
@@ -368,18 +395,28 @@ def initialize_customvalues
         end
 
         def initialize_scsi_controllers
-          if attributes[:scsi_controllers] && attributes[:scsi_controllers].is_a?(Array)
+          if attributes[:scsi_controllers] && attributes[:scsi_controllers].is_a?(Array) && !attributes[:scsi_controllers].empty?
             attributes[:scsi_controllers].map! do |controller|
               controller.is_a?(Hash) ? Fog::Vsphere::Compute::SCSIController.new(controller) : controller
             end
-          elsif attributes[:scsi_controller] && attributes[:scsi_controller].is_a?(Hash)
+          elsif attributes[:scsi_controller] && attributes[:scsi_controller].is_a?(Hash) && !attributes[:scsi_controller].empty?
+
             attributes[:scsi_controllers] = [
               Fog::Vsphere::Compute::SCSIController.new(attributes[:scsi_controller])
             ]
-          elsif attributes[:volumes] && attributes[:volumes].is_a?(Array) && !attributes[:volumes].empty?
-            # Create a default scsi controller if there are any disks but no controller defined
-            attributes[:scsi_controllers] = [
-              Fog::Vsphere::Compute::SCSIController.new
+          elsif unassigned_volumes? && !attributes[:volume].empty?
+            attributes[:scsi_controllers] = [Fog::Vsphere::Compute::SCSIController.new]
+          end
+        end
+
+        def initialize_nvme_controllers
+          if attributes[:nvme_controllers] && attributes[:nvme_controllers].is_a?(Array) && !attributes[:nvme_controllers].empty?
+            attributes[:nvme_controllers].map! do |controller|
+              controller.is_a?(Hash) ? Fog::Vsphere::Compute::NVMEController.new(controller) : controller
+            end
+          elsif attributes[:nvme_controller] && attributes[:nvme_controller].is_a?(Hash)
```

Comment:
Since we decided that `nvme_controller` should be removed, this part will need to be removed also.